### DISCUSSION TITLE
Proposal: Add TrOPD-style outlier FKL masking to GOLDTrainer

### DISCUSSION THREAD

**Discussion by Strongich** [huggingface/trl#5933] [Ideas]
Greetings!
A new paper on improving on-policy distillation was recently released: [Trust Region On-Policy Distillation](https://huggingface.co/papers/2606.01249).

I think one component from the paper could be a good fit for GOLDTrainer: outlier-aware FKL masking, which could improve the stability of the RKL / JSD path in a similar spirit to how DAPO-style loss variants improve stability in GRPOTrainer.

The motivation is the following. In RKL-based on-policy distillation, some student-generated tokens may receive extremely low probability under the teacher. These tokens can create very large negative token-level rewards / gradients and destabilize training. The paper addresses this by separating tokens into a **trust region** and an **outlier region**. For the trust-region tokens, it keeps the usual RKL-style objective. For outlier tokens, instead of applying the same RKL loss, it masks that RKL contribution and applies a top-k FKL loss on the teacher’s top-k distribution. In simplified form:

`loss = trust_region_mask * rkl_loss + outlier_mask * topk_fkl_loss`

This keeps the stabilizing effect of masking unreliable RKL gradients while still preserving some teacher signal for the outlier tokens, instead of discarding them entirely.

The reported results look promising. In the paper’s math-domain comparison, the FKL-outlier variant improves the average score from **46.79** for vanilla OPD to **49.00**. It also improves over generalized JSD with $\beta = 0.5$, which reports **47.90** in the same table. So this component alone gives a +2.21 improvement over OPD and +1.10 over JSD in that setting.
<img width="1440" height="531" alt="image" src="https://github.com/user-attachments/assets/ce586cd4-659f-4e98-8d39-e5d63c781ea3" />

The second component in the paper, Off-Policy Guidance, is also interesting, but I do not think it is the best first target for GOLDTrainer. It requires the teacher to generate an off-policy prefix, after which the student continues generation. This would introduce a teacher-generation path that GOLDTrainer does not currently rely on in the same way. It may also be expensive for long reasoning sequences: the paper starts with the teacher prefix length set to the maximum training sequence length and then anneals it down with a cosine schedule. For long OPD training, this could add substantial generation overhead and may complicate memory management, especially with vLLM / reserved-cache behavior.

My concrete proposal is:

1. Review the paper’s results and the implementation trade-offs for the outlier-FKL component.
2. When `use_uld_loss=False` and $\beta=1.0$, where the standard JSD path effectively becomes RKL-style distillation, enable an optional TrOPD-style outlier FKL objective by default, for example:
```
use_outlier_fkl_loss=True
outlier_fkl_top_k=64
```
3. For 0 < $\beta$ < 1, test this as an ablation by applying the outlier FKL correction only to the RKL-like part of the loss. Since the paper does not directly evaluate this mixed-JSD setting, I would keep it opt-in unless the ablation is clearly positive.
Keep Off-Policy Guidance out of the first implementation, unless maintainers think it is worth adding later as a separate feature.

If the Big HF Council approves this direction, I'll implement it, add tests, and run the relevant ablations.

**Comment by Strongich** [comment]
Update: I ran the A/B experiments -- backing away from the "make it default" part of this proposal.

## Setup

I implemented Eq. 5–7 (on-policy part only, as scoped above) in GOLDTrainer behind use_outlier_fkl_loss / outlier_fkl_top_k, and ran a controlled A/B on a VLM setting:

- Student: Qwen3-VL-2B-Instruct, Teacher: Qwen3-VL-8B-Instruct
- Data: GEOQA_R1V (5% held out for eval)
- Training: pure on-policy distillation (lmbda=1.0), beta=1.0 (reverse KL), temperature 0.6, vLLM colocate generation, LoRA (r=16) on the language model's q/k/v, lr 1e-4, effective batch 16, 500 steps
- The two runs differ only in use_outlier_fkl_loss (False = clean OPD baseline, True = TrOPD with top_k=64), same training seed

Evaluation: GEOQA held-out split (374 items, greedy decoding) + MathVista_MINI and MathVision_MINI via VLMEvalKit, 3 sampling seeds each (temperature 0.6, top_p 0.95), on the merged LoRA checkpoints.

## Results (500 steps)

| Benchmark | Base 2B (untrained) | Baseline (clean OPD) | TrOPD | Δ (TrOPD − baseline) |
|---|---:|---:|---:|---:|
| GEOQA holdout (374 items, greedy) | 33.96 | **37.43** | 36.63 | −0.80 |
| MathVista_MINI (3 seeds) | — | **52.20 ± 0.91** | 50.47 ± 0.25 | −1.73 |
| MathVision_MINI (3 seeds) | — | **20.50 ± 0.16** | 19.52 ± 1.09 | −0.98 |

A 100-step checkpoint showed the same picture (GEOQA 36.63% vs 34.49%, MathVista 51.33 vs 51.27, MathVision 20.18 vs 18.97). McNemar's test on the paired GEOQA predictions: no comparison is statistically significant (baseline vs TrOPD at 100 steps: p = 0.39; at 500 steps the net difference is 3 items, so p >= 0.25 under any pairing). So my read is no statistically significant effect, with the point estimates consistently on the wrong side -- TrOPD trails the plain-RKL baseline on all 3 benchmarks at both checkpoints.


<img width="1179" height="779" alt="image" src="https://github.com/user-attachments/assets/ee4a5b49-e4f5-44a7-8df9-78e461db33bd" />

The TrOPD train/loss is visibly lower, but the two curves are not comparable: train/outlier_token_frac (= expected fraction of tokens outside the trust region, Eq. 6) sits at ~12–14%, and on exactly those tokens TrOPD replaces the full-vocab reverse KL with a top-64 forward KL -- a term that is mechanically much smaller than what it replaces. So the lower loss reflects a changed objective, not a better student. 

## Revised suggestion

I no longer think this should become the default loss (my original DAPO-style framing). My config may of course be the problem -- one training seed, LoRA instead of full finetune, a fairly stable regime (temp 0.6, modest LR) where RKL may simply not produce the exploding-gradient outliers the paper targets. 

## Concretely, I'd propose:

1. Wait and see whether the paper gets accepted at a peer-reviewed venue and/or others reproduce the gains (possibly showing my config was the issue).
2. If so, land it as an opt-in flag (use_outlier_fkl_loss=False by default) — the implementation and A/B scripts are ready on my branch.
3. Definitely not the default for now.
